### PR TITLE
hash_ssz: simplify per latest spec update

### PR DESCRIPTION
**Pull request by arnetheduck**
* update per ethereum/eth2.0-specs#140
* join with ssz to form a single ssz module (like spec)

### CODE REVIEW

File: beacon_chain/ssz.nim [hunk 1]
```diff
@@ -12,7 +12,45 @@
 import ./datatypes, eth_common, endians, typetraits, options, nimcrypto
 
 # ################### Helper functions ###################################
-func `+`[T](p: ptr T, offset: int): ptr T {.inline.}=
+
+func len(x: Uint24): int = 3
+
+func toBytesSSZ(x: SomeInteger): array[sizeof(x), byte] =
+  ## Integers are all encoded as bigendian and not padded
```
Comment (mratsim) [inline beacon_chain/ssz.nim]: Not padded is ambiguous. Does that mean a varint

File: beacon_chain/ssz.nim [hunk 2]
```diff
@@ -12,7 +12,45 @@
...
+  let v = x.uint32
+  result[2] = byte(v and 0xff)
+  result[1] = byte((v shr 8) and 0xff)
+  result[0] = byte((v shr 16) and 0xff)
+
+func toBytesSSZ(x: EthAddress): array[sizeof(x), byte] = x
+func toBytesSSZ(x: MDigest[32*8]): array[32, byte] = x.data
+
+func fromBytesSSZUnsafe(T: typedesc, data: ptr byte): T =
```
Comment (mratsim) [inline beacon_chain/ssz.nim]: Is data an array of byte or an integer? In the second case the type should be changed to reflect that.

File: beacon_chain/ssz.nim [hunk 3]
```diff
@@ -12,7 +12,45 @@
...
+  ## Integers are all encoded as bigendian and not padded
+  ## Assumes no buffer overruns!
+
+  # XXX: any better way to get a suitably aligned buffer in nim???
```
Comment (mratsim) [inline beacon_chain/ssz.nim]: Type-defined or allocator-defined alignment is an aim for v1 destructors.

See: https://github.com/nim-lang/Nim/wiki/Destructors#allocators
```Nim
type
  Allocator* {.inheritable.} = ptr object
    alloc*: proc (a: Allocator; size: int; alignment = 8): pointer {.nimcall.}
    dealloc*: proc (a: Allocator; p: pointer; size: int) {.nimcall.}
    realloc*: proc (a: Allocator; p: pointer; oldSize, newSize: int): pointer {.nimcall.}
```

In the mean time, this is my alignment proc that I use for SIMD vectorization in my tensor libraries: https://github.com/numforge/laser/blob/master/laser/private/memory.nim#L10

```Nim
withCompilerOptimHints()
  # Defines
  #   - LASER_MEM_ALIGN = 64
  #   - assume_aligned
  #   - and {.restrict.}
  # which are optimization hints that are not needed for Nimbus.

func align_raw_data*(T: typedesc, p: pointer): ptr UncheckedArray[T] =
  static: assert T.supportsCopyMem

  let address = cast[ByteAddress](p)
  let aligned_ptr{.restrict.} = block: # We cannot directly apply restrict to the default "result"
    let remainder = address and (LASER_MEM_ALIGN - 1) # modulo LASER_MEM_ALIGN (power of 2)
    if remainder == 0:
      assume_aligned cast[ptr UncheckedArray[T]](address)
    else:
      let offset = LASER_MEM_ALIGN - remainder
      assume_aligned cast[ptr UncheckedArray[T]](address +% offset)
  return aligned_ptr
```

File: beacon_chain/ssz.nim (hunk 2 above)
Comment (mratsim) [reply beacon_chain/ssz.nim]: Actually maybe not but we need a
`static: assert T.sizeof <= 8` as we use an uint64 for alignment.

Comment (arnetheduck) [reply beacon_chain/ssz.nim]: :+1:  

already handled below by `when result.sizeof...` and `{.fatal.}`, however, adding `T: SomeInteger` makes this intention more clear.

I'd actually prefer using an `openarray[byte]` for `data`, but that would take more surgery on the API - will probably come back to it later - just waiting to see if the little-endian proposal will pass.

File: beacon_chain/ssz.nim (hunk 3 above)
Comment (arnetheduck) [reply beacon_chain/ssz.nim]: yeah, it's an old `C` trick, from way back before https://en.cppreference.com/w/cpp/language/alignas :) 

alignment support generally needs to address two things - what you want from the compiler and what the compiler can expect from you: ie for types where you tell the compiler what alignment it should use, and for variables/pointers where you tell it what alignment promises you're making, for inputs (similar to how restrict makes promises to the compiler about aliasing).. 

I guess a third thing might be the underlying memory allocator and convenience api's in it to support things - basically what those nim destructors / allocators are about, but that's just details. 

Oh, and as usual, whatever alignment/sizeof support there is in nim, it's broken :) https://github.com/nim-lang/Nim/issues/9667

File: beacon_chain/ssz.nim (hunk 1 above)
Comment (arnetheduck) [reply beacon_chain/ssz.nim]: pushed the spec text instead, now there's someone else to blame :)

**Pull request merged**